src/compiler/program.ts

```diff
@@ -2984,6 +3072,35 @@ namespace ts {
             }
         }
 
+        function addProgramDiagnosticAtRefPath(file: SourceFile, rootPaths: Map<true>, message: DiagnosticMessage, ...args: (string | number | undefined)[]) {
+            const refPaths = refFileMap && refFileMap.get(file.path);
+            const refPathToReportErrorOn = forEach(refPaths, refPath => rootPaths.has(refPath.file) ? refPath : undefined) ||
+                elementAt(refPaths, 0);
+            if (refPathToReportErrorOn) {
+                const refFile = Debug.assertDefined(getSourceFileByPath(refPathToReportErrorOn.file));
+                const { kind, index } = refPathToReportErrorOn;
+                let pos: number, end: number;
+                switch (kind) {
+                    case RefFileKind.Import:
+                        pos = skipTrivia(refFile.text, refFile.imports[index].pos);
+                        end = refFile.imports[index].end;
+                        break;
+                    case RefFileKind.ReferenceFile:
+                        ({ pos, end } = refFile.referencedFiles[index]);
+                        break;
+                    case RefFileKind.TypeReferenceDirective:
+                        ({ pos, end } = refFile.typeReferenceDirectives[index]);
+                        break;
+                    default:
+                        return Debug.assertNever(kind);
+                }
+                programDiagnostics.add(createFileDiagnostic(refFile, pos, end - pos, message, ...args));
+            }
+            else {
+                programDiagnostics.add(createCompilerDiagnostic(message, ...args));

```

Comment:
looks like this happens when refFileMap's entry for `file` is empty. When does that happen?

Reply:
This is possible if error to report happens to be in rootPath? Eg checkout test in `unittests/tsserver/telemetry.ts` called `does not expose paths` The root file `/a.ts` does not belong to `rootDir`: `/hunter2`